language/tools/move-cli/src/main.rs

```diff
@@ -0,0 +1,440 @@
+    assert!(effects.modules.is_empty());
+    if !effects.events.is_empty() {
+        println!("Emitted {:?} events:", effects.events.len());
+        // TODO: better event printing
+        for (event_handle, event_sequence_number, _event_type, _event_layout, event_data) in
+            &effects.events
+        {
+            println!(
+                "Emitted {:?} as the {}th event to stream {:?}",
+                event_data, event_sequence_number, event_handle
+            )
+        }
+        // TODO: support saving events to disk
+        println!("Warning: saving events to disk is currently not supported. Discarding events.");
+    }
+    if !effects.resources.is_empty() {
+        println!(
+            "Changed resource(s) under {:?} address(es):",
+            effects.resources.len()
+        );
+    }
+    for (addr, writes) in &effects.resources {
+        print!("  ");
+        println!(
+            "Changed {:?} resource(s) under address {:?}:",
+            writes.len(),
+            addr
+        );
+        for (struct_tag, write_opt) in writes {
+            print!("    ");
+            match write_opt {
+                Some((_layout, value)) => {
+                    if state
+                        .get_resource_bytes(*addr, struct_tag.clone())?
+                        .is_some()
+                    {
+                        // TODO: print resource diff
+                        println!("Changed type {}: {}", struct_tag, value)
+                    } else {
+                        // TODO: nicer printing
+                        println!("Added type {}: {}", struct_tag, value)
+                    }
+                }
+                None => println!("Deleted type {}", struct_tag),
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Commit the resources and modules modified by a transaction to disk
+fn maybe_commit_effects(
+    commit: bool,
+    effects_opt: Option<TransactionEffects>,
+    state: &OnDiskStateView,
+) -> Result<()> {
+    if commit {
+        if let Some(effects) = effects_opt {
+            for (addr, writes) in effects.resources {
+                for (struct_tag, write_opt) in writes {
+                    match write_opt {
+                        Some((layout, value)) => {
+                            state.save_resource(addr, struct_tag, layout, value)?
+                        }
+                        None => state.delete_resource(addr, struct_tag)?,
+                    }
+                }
+            }
+        }
+        // TODO: print modules to be saved?
+        state.save_modules()?;
+        println!("Committed changes.")
+    } else if !effects_opt.map_or(false, |effects| effects.resources.is_empty()) {
+        println!("Discarding changes; re-run with --commit if you would like to keep them.")
+    }
+
+    Ok(())
+}
+
+/// Explain an execution error
+fn explain_error(error: VMError, state: &OnDiskStateView) -> Result<()> {
```

Comment:
woah....

Some neat stuff. We should probably move this to VMStatus maybe? (bit tricky with the OnDiskStateView as is now) but might be useful for other tools

Reply:
Agreed that putting this somewhere more generally usable would be good! Not sure about `VMStatus` for the reason you mention + that it would be nice to show source context as part of this too. 

FWIW most of this is building on @tzakian's awesome work with the error map.